File: 3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii/3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii.js
/**
 * @param {string} word
 * @param {number} k
 * @return {number}
 */
function countOfSubstrings(word, k) {
  const n = word.length;

  // Build a prefix sum array where prefixConsonantCount[i] represents
  // the number of consonants in word[0...i-1].
  const prefixConsonantCount = new Int32Array(n + 1);

  // Create a fast lookup for vowels (using ASCII codes)
  const isVowel = new Uint8Array(128);
  isVowel['a'.charCodeAt(0)] = 1;
  isVowel['e'.charCodeAt(0)] = 1;
  isVowel['i'.charCodeAt(0)] = 1;
  isVowel['o'.charCodeAt(0)] = 1;
  isVowel['u'.charCodeAt(0)] = 1;

  for (let i = 0; i < n; i++) {
    const charCode = word.charCodeAt(i);
    // If vowel, add 0; otherwise, add 1 to count a consonant.
    prefixConsonantCount[i + 1] = prefixConsonantCount[i] + (isVowel[charCode] ? 0 : 1);
  }

  // Frequency array to count occurrences of prefix sums for valid starting indices.
  const prefixFrequency = new Uint32Array(n + 1);
  let validStartPointer = 0; // Pointer for valid starting indices
  let validSubstringCount = 0;

  // Track last occurrence of each vowel. Initialized to -1 (not seen).
  let lastA = -1, lastE = -1, lastI = -1, lastO = -1, lastU = -1;

  // Iterate over each possible ending index of a substring.
  for (let rightIndex = 0; rightIndex < n; rightIndex++) {
    const currentCharCode = word.charCodeAt(rightIndex);
    // Update last seen position for vowels.
    if (currentCharCode === 97) { // 'a'
      lastA = rightIndex;
    } else if (currentCharCode === 101) { // 'e'
      lastE = rightIndex;
    } else if (currentCharCode === 105) { // 'i'
      lastI = rightIndex;
    } else if (currentCharCode === 111) { // 'o'
      lastO = rightIndex;
    } else if (currentCharCode === 117) { // 'u'
      lastU = rightIndex;
    }

    // If not all vowels have been seen yet, skip this ending index.
    if (lastA === -1 || lastE === -1 || lastI === -1 || lastO === -1 || lastU === -1) {
      continue;
    }

    // The valid starting indices for substrings ending at rightIndex must be ≤
    // the smallest (earliest) last occurrence among the vowels.
    const minValidStartIndex = Math.min(lastA, lastE, lastI, lastO, lastU);

    // Activate all starting indices up to minValidStartIndex by updating their prefix sum counts.
    while (validStartPointer <= minValidStartIndex) {
      prefixFrequency[prefixConsonantCount[validStartPointer]]++;
      validStartPointer++;
    }

    // For a substring from L to rightIndex to have exactly k consonants:
    // prefixConsonantCount[rightIndex + 1] - prefixConsonantCount[L] === k,
    // which implies prefixConsonantCount[L] === prefixConsonantCount[rightIndex + 1] - k.
    const targetPrefixSum = prefixConsonantCount[rightIndex + 1] - k;
    if (targetPrefixSum >= 0 && targetPrefixSum <= n) {
      validSubstringCount += prefixFrequency[targetPrefixSum];
    }
  }

  return validSubstringCount;
}

File: 3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii/3569-count-of-substrings-containing-every-vowel-and-k-consonants-ii.java
class Solution {
    private long countOfAtleastKSubstrings(String word, int k){
        long result = 0;
        long n = word.length();
        long a = 0, e = 0, i = 0, o = 0, u = 0, consonants = 0;
        for(int right = 0, left = 0;  right < n; right++){
            switch(word.charAt(right)){
                case 'a':
                    a++;
                    break;
                case 'e':
                    e++;
                    break;
                case 'i':
                    i++;
                    break;
                case 'o':
                    o++;
                    break;
                case 'u':
                    u++;
                    break;
                default:
                    consonants++;
                    break;
            }
            while(a > 0 && e > 0 &&  i > 0 && o > 0 && u > 0 && consonants >= k){
                switch(word.charAt(left)){
                    case 'a':
                        a--;
                        break;
                    case 'e':
                        e--;
                        break;
                    case 'i':
                        i--;
                        break;
                    case 'o':
                        o--;
                        break;
                    case 'u':
                        u--;
                        break;
                    default:
                        consonants--;
                        break;
                }
                result += (n - right);
                left++;
            }
        }
        return result;
    }

    public long countOfSubstrings(String word, int k) {
        return countOfAtleastKSubstrings(word, k) - countOfAtleastKSubstrings(word, k+1);
    }
}